[SE-0206][stdlib] Implement Hashable Enhancements proposal

lorentey opened the pull request:
This PR implements the changes proposed in [SE-0206], "Hashable Enhancements".

* Makes the `Hasher` type and `Hashable`'s new `hash(into:)` requirement public
* Extends compiler support for deriving `Hashable` with new cases, as described below.

This is fully source compatible with existing code. Code can choose to implement either Hashable requirement; the other requirement is then automatically synthesized by the compiler.

rdar://problem/35052153

## Conforming to `Hashable` by implementing `hashValue`

Code written for previous versions of Swift conforms to `Hashable` by implementing `hashValue`. This will keep working after this PR, with no source-level changes:

```swift
struct OldFashionedPoint: Hashable {
  let x: Int
  let y: Int
  var hashValue: Int {
    return x.hashValue ^ y.hashValue &* 16777619
  }
}
```
To make this work, the compiler synthesizes the missing `hash(into:)` requirement automatically:
```swift
  @derived func hash(into hasher: inout Hasher) {
    hasher.combine(self.hashValue)
  }
```

This partial synthesis is new with this PR. It works for all types that can implement `Hashable`: enums, structs and classes, with no restrictions. (Deriving class members is now safe, thanks to @slavapestov's work in #16057.)

## Conforming to `Hashable` by implementing `hash(into:)`

I hope code written for 4.2+ will prefer to implement the `hash(into:)` requirement instead:
```swift
struct ModernPoint: Hashable {
  let x: Int
  let y: Int
  func hash(into hasher: inout Hasher) {
    hasher.combine(x)
    hasher.combine(y)
  }
}
```
In this case the compiler will automatically provide the missing `hashValue` requirement, with the following implementation:
```swift
  @derived var hashValue: Int {
    var hasher = Hasher()
    self.hash(into: &hasher)
    return hasher.finalize()
  }
```

Like the previous case, this kind of `hashValue` synthesis works for all types, with no restrictions.


## Conforming to `Hashable` by automatic synthesis

Since [SE-0185], it has been possible to have the compiler supply the entire `Hashable` conformance for structs and enums whose components are all `Hashable`. For example:

```swift
struct Point: Hashable {
  let x: Int
  let y: Int
}
```
This will keep working after this PR, except the compiler now implements this conformance as follows:
```swift
  @derived func hash(into hasher: inout Hasher) {
    hasher.combine(x)
    hasher.combine(y)
  }
  @derived var hashValue: Int { // Same as above
    var hasher = Hasher()
    self.hash(into: &hasher)
    return hasher.finalize()
  }
```

Enums work, too:
```swift
enum Shape: Hashable {
  case point
  case circle(radius: Int)
  case rectangle(width: Int, height: Int)

  @derived func hash(into hasher: inout Hasher) {
    switch self {
    case .point:
      hasher.combine(0)
    case .circle(radius: let radius):
      hasher.combine(1)
      hasher.combine(radius)
    case .rectangle(width: let width, height: let height):
      hasher.combine(2)
      hasher.combine(width)
      hasher.combine(height)
    }
  }
  @derived var hashValue: Int {
    var hasher = Hasher()
    self.hash(into: &hasher)
    return hasher.finalize()
  }
}
```

(In the examples above, the compiler also synthesizes `Equatable` conformance.)

[SE-0206]: https://github.com/apple/swift-evolution/blob/master/proposals/0206-hashable-enhancements.md
[SE-0185]: https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md

lorentey committed:
[SE-0206][Sema] Update Hashable synthesizer to support hash(into:)

This removes the default implementation of hash(into:), and replaces it with automatic synthesis built into the compiler. Hashable can now be implemented by defining either hashValue or hash(into:) -- the compiler supplies the missing half automatically, in all cases.

To determine which hash(into:) implementation to generate, the synthesizer resolves hashValue -- if it finds a synthesized definition for it, then the generated hash(into:) body implements hashing from scratch, feeding components into the hasher. Otherwise, the body implements hash(into:) in terms of hashValue.

lorentey committed:
[SE-0206][CoreFoundation] _CFObject: Provide default implementation for _hash(into:)

Without this change, the Hashable synthesizer attempts to add hash(into:) methods directly on CF types, which (somewhat unsurprisingly) fails with assertion below. (This situation is unique to CF types, which are imported with an implicit _CFObject conformance; we usually have an extension context or a non-imported type decl in which to put the derived methods.)

Assertion failed: (!decl->isForeign() && "Use getForeignMetadataLayout()"), function getClassMetadataLayout, file /Users/klorentey/Swift/swift/lib/IRGen/MetadataLayout.cpp, line 83.
0  swift                    0x000000010ccd29c8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x000000010ccd30d6 SignalHandler(int) + 694
2  libsystem_platform.dylib 0x00007fff600e0d9a _sigtramp + 26
3  swift                    0x000000010e04f1ed cmark_strbuf__initbuf + 122440
4  libsystem_c.dylib        0x00007fff5ffa6f79 abort + 127
5  libsystem_c.dylib        0x00007fff5ff6f090 basename_r + 0
6  swift                    0x00000001096e8f95 swift::irgen::IRGenModule::getClassMetadataLayout(swift::ClassDecl*) + 53
7  swift                    0x00000001095b9e25 swift::irgen::emitVirtualMethodValue(swift::irgen::IRGenFunction&, llvm::Value*, swift::SILDeclRef, swift::CanTypeWrapper<swift::SILFunctionType>) + 133
8  swift                    0x00000001095ba252 swift::irgen::emitVirtualMethodValue(swift::irgen::IRGenFunction&, llvm::Value*, swift::SILType, swift::SILDeclRef, swift::CanTypeWrapper<swift::SILFunctionType>, bool) + 690
9  swift                    0x00000001096bef49 swift::SILInstructionVisitor<(anonymous namespace)::IRGenSILFunction, void>::visit(swift::SILInstruction*) + 33497
10 swift                    0x00000001096b3067 swift::irgen::IRGenModule::emitSILFunction(swift::SILFunction*) + 8055
11 swift                    0x00000001095c8d8b swift::irgen::IRGenerator::emitLazyDefinitions() + 1051
12 swift                    0x000000010968eb16 performIRGeneration(swift::IRGenOptions&, swift::ModuleDecl*, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, swift::SourceFile*, llvm::GlobalVariable**, unsigned int) + 1382
13 swift                    0x000000010968f05e swift::performIRGeneration(swift::IRGenOptions&, swift::SourceFile&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, unsigned int, llvm::GlobalVariable**) + 94
14 swift                    0x000000010952cfc0 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 15488
15 swift                    0x0000000109528332 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2962
16 swift                    0x00000001094e4848 main + 2328
17 libdyld.dylib            0x00007fff5feecc41 start + 1
Stack dump:
[…]
1.	While emitting IR SIL function "@$SSo10CGColorRefas8HashableSCsACP4hash4intoys6HasherVz_tFTW".
 for 'hash(into:)' in module 'CoreGraphics'

lorentey committed:
[SE-0206][stdlib] De-underscore Hasher

lorentey committed:
[SE-0206][stdlib] Document Hasher

lorentey commented on stdlib/public/core/Hasher.swift:
```diff
@@ -331,16 +360,22 @@ public struct _Hasher {
 
+  /// Finalize the hasher state and return the hash value.
+  /// Finalizing invalidates the hasher; additional bits cannot be combined
+  /// into it, and it cannot be finalized again.
```
You're right, it isn't! I have to change it to be nonmutating but consuming. I don't think we need to change the name to `finalized()` because of the consuming semantics. @jckarter, do you agree?

jckarter replied:
We didn't discuss changing the name. The discussion about the hasher being invalidated may still not be necessary anymore since the consuming means you can't use the finalized value anymore anyway.

lorentey committed:
[SE-0206][stdlib] Make Hasher.finalize() nonmutating but __consuming

As noted in the proposal’s revision, this allows us to get rid of finalization checks, improves API robustness, and paves the way for making Hasher move-only in the future.

lorentey committed:
[lit] Add %target-run-stdlib-swiftgyb

lorentey committed:
[SE-0206][stdlib] Hasher, SipHash: Make non-public declarations internal

Newly internal declarations include Hasher._seed and the integer overloads of Hasher._combine(_:), as well as _SipHash13 and _SipHash24.

Unify the interfaces of these SipHash testing structs with Hasher. Update SipHash test to cover Hasher, too.

Add @usableFromInline to all newly internal stuff. In addition to its normal use, it also enables white box testing; compile tests that need to use these declarations with -disable-access-control.

lorentey commented:
OK, we now have a consuming finalize. I also made non-public declarations `@usableFromInline internal`; this still allows white-box testing with `-disable-access-control`, but it makes them much trickier to access from user code.

lorentey commented:
@swift-ci please test

lorentey commented on stdlib/public/core/Hasher.swift:
```diff
@@ -377,7 +356,16 @@ public struct Hasher {
+  /// Finalize the hasher state and return the hash value. Finalizing consumes
+  /// the hasher, forbidding further operations.
+  @effects(releasenone)
+  public __consuming func finalize() -> Int {
```
@jckarter, @natecook1000 A later commit changes it to read "Finalizing consumes the hasher, forbidding further operations", which is technically kind of true, but it's not the whole story. How about this?

```swift
public Hasher {
  ...
  /// Finalize the hasher state and return the hash value.
  ///
  /// Finalizing consumes the hasher: it is illegal to finalize a hasher you
  /// don't own, or to perform operations on a finalized hasher. (These
  /// may become compile-time errors in the future.)
  @effects(releasenone)
  public __consuming func finalize() -> Int {...}
}
...
public protocol Hashable {
  ...
  /// Hash the essential components of this value into the hash function
  /// represented by `hasher`, by feeding them into it using its `combine`
  /// methods.
  ///
  /// Essential components are precisely those that are compared in the type's
  /// implementation of `Equatable`.
  ///
  /// Note that `hash(into:)` doesn't own the hasher passed into it, so it must not
  /// call `finalize()` on it. Doing so may become a compile-time error in the future.
  func hash(into hasher: inout Hasher)
}

```

lorentey committed:
[SE-0206][stdlib] Clarify Hasher.finalize() documentation

lorentey committed:
[SE-0206][stdlib] De-underscore Hashable.hash(into:)

lorentey committed:
[SE-0206][test] Fix tests that look at Hashable requirements

hash(into:) needs to be included in expectations; tests looking at synthesized Hashable implementation bodies need to be updated for resilient hashing.

lorentey committed:
[SE-0206][Sema] Add separate hash(into:) impl for enums with no associated values

For enums with no associated values, it is better to move the hasher.combine call out of the switch statement, like this:

func hash(into hasher: inout Hasher) {
  let discriminator: Int
  switch self {
    case a: discriminator = 0
    case b: discriminator = 1
    case c: discriminator = 2
  }
  hasher.combine(discriminator)
}

This enables the optimizer to replace the switch statement with a simple integer promotion, restoring earlier behavior.

lorentey committed:
[SE-0206][test] Update diagnostic tests for Hashable derivation

lorentey commented:
I'll have to quickly rebase this to incorporate the new StdlibUnittest functionality from #16066.

Then I'll cherry-pick & update commits from #15122 so this PR will implement the whole of SE-0206, including de-underscoring `hash(into:)` and extending compiler synthesis.

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test source compatibility

lorentey commented:
@swift-ci please test source compatibility

lorentey commented:
cc @slavapestov 356e595a86c3528ccefdb2da965dc1153b2350ad and ad1255b24191b22f2a0b67df8b4ca8c1c1f21638 implement synthesis for the new Hashable.

Adding `Hashable` in `finalizeType()` fixed the vtable issue with synthesized class methods; thank you!

lorentey commented on stdlib/public/SDK/CoreFoundation/CoreFoundation.swift:
```diff
@@ -17,6 +17,9 @@ extension _CFObject {
   }
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.hashValue)
+  }
```
CF types are magically auto-comformed to the `_CFObject` protocol, and synthesis doesn't work for them. Luckily this is an extremely special use case that (as far as I know) only applies to this particular protocol. (It definitely doesn't affect any user code.) cc @jrose-apple

lorentey committed:
[stdlib] Remove obsolete hashing interfaces

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
@swift-ci smoke test compiler performance

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
@swift-ci smoke test compiler performance

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

milseman reviewed (approved):
LGTM, though you'll want to loop in someone to review the code synthesis part.
